poc/observe.py

```diff
@@ -0,0 +1,764 @@
+    callback : callable(object, name, old, new)
+        A callable conforming to the notifier signature.
+    path : ListenerPath
+        Path for listening to extended traits.
+    remove : boolean
+        Whether to remove the observer.
+    dispatcher : callable(callable, args, kwargs)
+        Callable for dispatching the callback, i.e. dispatching
+        callback on a different thread.
+    """
+    listener = path.node
+    for this_target in listener.iter_this_targets(object):
+        if listener.notify:
+            notifier = TraitObserverNotifier(
+                observer=callback,
+                owner=this_target._notifiers(True),
+                target=target,
+                event_factory=listener.event_factory,
+                dispatcher=dispatcher,
+            )
+            remove_notifier(this_target, notifier)
+
+        for next_path in path.nexts:
+
+            change_notifier = listener.get_change_notifier(
+                callback=callback,
+                path=next_path,
+                target=target,
+                dispatcher=dispatcher,
+            )
+            remove_notifier(object=this_target, notifier=change_notifier)
+
+            for next_target in listener.iter_next_targets(object):
+                remove_notifiers(
+                    object=next_target,
+                    callback=callback,
+                    path=next_path,
+                    target=target,
+                    dispatcher=dispatcher,
+                )
+
+
+def is_notifiable(object):
+    """ Return true if an object conforms to the expected
+    interface for a notifiable object.
+    """
+    return isinstance(object, INotifiableObject)
+
+
+class BaseListener:
+
+    def event_factory(self, object, name, old, new):
+        raise NotImplementedError()
+
+    def __eq__(self, other):
+        """ Return true if a given instance is equivalent to this
+        one. Needed for comparing paths and cleaning up listeners.
+        """
+        raise NotImplementedError()
+
+    @property
+    def notify(self):
+        """ Whether to call notifiers for changes on this item."""
+        return getattr(self, "_notify", True)
+
+    @notify.setter
+    def notify(self, value):
+        self._notify = value
+
+    def iter_this_targets(self, object):
+        """ Yield (notifiable) objects for attaching notifiers for this
+        listener.
+        """
+        yield from ()
+
+    def iter_next_targets(self, object):
+        """ Yield (notifiable) objects for attaching notifiers for the
+        next listener following this one in a ListenerPath.
+        """
+        # For walking down the path of Listeners
+        yield from ()
+
+    def get_change_notifier(self, callback, path, target, dispatcher):
+        """ Return a ListenerChangeNotifier for removing/propagating listeners
+        for the 'next' targets.
+        """
+        raise NotImplementedError()
+
+
+class AnyTraitListener(BaseListener):
```

Comment:
This might need an `optional` flag just like the named trait case. Eg. in the current notation, handling something like `+:foo`.